### PR TITLE
:sparkles: Add the grouping loss to single model training

### PR DESCRIPTION
**Pull request by o-laurent**
To do:

- [x] Add tests

### CODE REVIEW

File: torch_uncertainty/routines/classification.py [hunk 1]
```diff
@@ -172,6 +177,13 @@ def __init__(
             self.test_ood_metrics = ood_metrics.clone(prefix="hp/test_")
             self.test_entropy_ood = Entropy()
 
+        if self.eval_grouping_loss:
+            grouping_loss = MetricCollection(
```
Comment (o-laurent) [inline torch_uncertainty/routines/classification.py:181]: This is probably a bit useless, although friendlier in the test step.

File: torch_uncertainty/routines/classification.py [hunk 2]
```diff
@@ -205,8 +217,14 @@ def criterion(self) -> nn.Module:
             self.loss = partial(self.loss, model=self.model)
         return self.loss()
 
-    def forward(self, inputs: Tensor) -> Tensor:
-        return self.model.forward(inputs)
+    def forward(self, inputs: Tensor, return_features: bool = False) -> Tensor:
```
Comment (o-laurent) [inline torch_uncertainty/routines/classification.py:220]: What do you think about this part @alafage?

File: torch_uncertainty/routines/classification.py [hunk 3]
```diff
@@ -595,6 +644,12 @@ def __init__(
                 prefix="hp/test_ood_ens_"
             )
 
+        if self.eval_grouping_loss:
```
Comment (o-laurent) [inline torch_uncertainty/routines/classification.py:647]: you need feat_forward for the ensembles, that we don't have yet

File: torch_uncertainty/routines/classification.py [hunk 4]
```diff
@@ -205,8 +217,14 @@ def criterion(self) -> nn.Module:
             self.loss = partial(self.loss, model=self.model)
         return self.loss()
 
-    def forward(self, inputs: Tensor) -> Tensor:
-        return self.model.forward(inputs)
+    def forward(self, inputs: Tensor, return_features: bool = False) -> Tensor:
+        if return_features:
+            features = self.model.feats_forward(inputs)
+            logits = self.model.linear(features)
```
Comment (o-laurent) [inline torch_uncertainty/routines/classification.py:223]: the last layer must be implicitly called "linear", we could also ask for a "classification_head" nn.Module in networks when we want to compute the grouping loss

File: torch_uncertainty/routines/classification.py [hunk 5]
```diff
@@ -453,6 +493,11 @@ def add_model_specific_args(
         parent_parser.add_argument(
             "--logits", dest="use_logits", action="store_true"
         )
+        parent_parser.add_argument(
```
Comment (o-laurent) [inline torch_uncertainty/routines/classification.py:496]: Actually, I found it quite fast so the main issue with the GL is that we need the models to have a feats_forward maybe

File: torch_uncertainty/routines/classification.py (hunk 2 above)
Comment (alafage) [reply torch_uncertainty/routines/classification.py:220]: I think it's a clean and efficient way to do it! Don't forget to update the signature accordingly:
```py
-> tuple[Tensor, Tensor | None]:
```

File: torch_uncertainty/routines/classification.py (hunk 5 above)
Comment (alafage) [reply torch_uncertainty/routines/classification.py:496]: Is that something we could check beforehand? To make sure nobody asks for the grouping_loss without the `feats_foward()` method in their model?

Comment (o-laurent) [reply torch_uncertainty/routines/classification.py:496]: We can check it with hasattr at the beginning of the routine perhaps?

File: torch_uncertainty/routines/classification.py (hunk 2 above)
Comment (o-laurent) [reply torch_uncertainty/routines/classification.py:220]: Yes, thx!
